### PR TITLE
Generate the first SBOM protoype from the Kubernetes release process

### PR DESCRIPTION
**Pull request by puerco**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

This PR integrates the Bill of Materials code Into the release process. This generates the first SBOM prototypes that describe a Kubernetes release. We are generating one SBOM for the source code and one for the release artifacts. The proposed namespaces for these are the following:

* http://k8s.io/sbom/source/v1.22.0-alpha.3
* http://k8s.io/sbom/release/v1.22.0-alpha.3

This PR aims to test the integration of the BOM code with the Kubernetes Release process. The resulting SBOM documents will describe the release source and artifacts, but will not include yet all the features of the SPDX libraries. They will do most of the work of creating the BOM however, enabling us to look for bugs and polish the documents to their final state in beta.0.

This PR is broken into several commits, I've attempted to add meaningful descriptions to each one. The most important changes are:

* Prewarming the license cache during release.PrepareWorkspaceStage() to avoid downloading the license files later
* The release process staging phase now has a `GenerateBillOfMaterials()` step that builds the SPDX documents.
* We now create an SPDX SBOM describing the Kubernetes source during staging
* Each version in a release now features an SPDX bill of materials listing its binaries and images
* stage.GenerateBillOfMaterials() now has an integration test

#### Which issue(s) this PR fixes:
Part of: https://github.com/kubernetes/release/issues/1837

#### Special notes for your reviewer:

The SBOM generation has been tested in several test runs of the release process. I've verified that both stage and release work as expected. For details see:

* Mock stage run: https://console.cloud.google.com/cloud-build/builds/c95c4cde-bfa6-470f-ab35-540399f90a0c?project=648026197307
* Mock release run: https://console.cloud.google.com/cloud-build/builds/82f86e12-7a0c-492d-93ef-d1a8b64f9f41?project=648026197307

One of the commits features a temporary function: `ListArtifacts()`. This function will be replaced with the Supported Platforms framework once we finish it  (ref: https://github.com/kubernetes/release/pull/1926 )

#### Does this PR introduce a user-facing change?

```release-note
* When staging a new kubernetes build, `krel` will now prewarm the license cache to have the classifier data ready when generating the bill of materials.
* The release process staging phase now has a `GenerateBillOfMaterials()` step that builds the SPDX documents.
* We now create an SPDX SBOM describing the Kubernetes source during staging
* Each version in a release now features an SPDX bill of materials listing its binaries and images
* stage.GenerateBillOfMaterials() now has an integration test

```

### CODE REVIEW

**Comment by puerco** [comment]
I gave it a test run with `export TOOL_ORG=puerco && export TOOL_REF=bom-rproc-1 ` (using a slightly modified version as currently there is [a problem with the release process](#2094)) and it works:
```
[2021-05-29T06:02:24Z] Found release versions: prime: v1.22.0-alpha.3, alpha: v1.22.0-alpha.3 
INFO[2021-05-29T06:02:24Z] Preparing workspace                           step=5/9
INFO[2021-05-29T06:02:24Z] Caching SPDX license set to /tmp.k8s/spdx/downloadCache 
INFO[2021-05-29T06:02:24Z] Downloading main SPDX license data from https://spdx.org/licenses/ 
INFO[2021-05-29T06:02:24Z] Sending GET request to https://spdx.org/licenses/licenses.json 
INFO[2021-05-29T06:02:25Z] Read data for 460 licenses. Downloading.     
INFO[2021-05-29T06:02:25Z] Downloading license data from https://spdx.org/licenses/UCL-1.0.json 
...
INFO[2021-05-29T06:02:31Z] Downloaded 460 licenses                      
INFO[2021-05-29T06:02:31Z] Preparing workspace for staging in /workspace/src/k8s.io/kubernetes 
INFO[2021-05-29T06:02:31Z] Cloning repository to /workspace/src/k8s.io/kubernetes 
```

**Comment by puerco** [comment]
Since this PR has not merged I will further this branch to generate the first BOM prototypes from the release process.
/hold

File: pkg/anago/stage.go [hunk 1]
```diff
@@ -500,6 +626,33 @@ func (d *DefaultStage) GenerateChangelog() error {
 	})
 }
 
+func (d *DefaultStage) GenerateBillOfMaterials() error {
+	for _, version := range d.state.versions.Ordered() {
+		artifactsList, err := d.impl.ListArtifacts(version)
+		if err != nil {
+			return errors.Wrap(err, "getting artifacts list")
+		}
+		logrus.Infof("Artifacts list: %+v", artifactsList)
```
Comment (saschagrunert) [inline pkg/anago/stage.go:635]: I'm not sure if we need this. Maybe we can list them directly during iteration to be better readable.

File: pkg/license/download.go [hunk 2]
```diff
@@ -60,7 +60,7 @@ func NewDownloaderWithOptions(opts *DownloaderOptions) (*Downloader, error) {
 type DownloaderOptions struct {
 	EnableCache       bool   // Should we use the cache or not
 	CacheDir          string // Directory where data will be cached, defaults to temporary dir
-	parallelDownloads int    // Number of license downloads we'll do at once
+	ParallelDownloads int    // Number of license downloads we'll do at once
```
Comment (saschagrunert) [inline pkg/license/download.go:63]: Do we need to make this public?

File: pkg/anago/stage.go [hunk 3]
```diff
@@ -225,6 +232,125 @@ func (d *defaultStageImpl) GenerateChangelog(options *changelog.Options) error {
...
 }
 
+func (d *defaultStageImpl) GenerateVersionArtifactsBOM(options *spdx.DocGenerateOptions) error {
+	_, err := spdx.NewDocBuilder().Generate(options)
+	return errors.Wrap(err, "generating bill of materials")
+}
+
+// ListArtifacts is a function lists the release artifacts, will be
+// replaced once the supported platforms code is ready
+func (d *defaultStageImpl) ListArtifacts(version string) (list map[string][]string, err error) {
```
Comment (saschagrunert) [inline pkg/anago/stage.go:242]: Note: As discussed with @puerco, this is a temporarily function which will be moved to a different place later on.

File: pkg/anago/anago.go [hunk 4]
```diff
@@ -279,6 +279,11 @@ func (s *Stage) Run() error {
 		return errors.Wrap(err, "generate changelog")
 	}
 
+	logger.WithStep().Info("Generating bill of materials")
+	if err := s.client.GenerateBillOfMaterials(); err != nil {
+		return errors.Wrap(err, "generating sbom")
+	}
```
Comment (saschagrunert) [inline pkg/anago/anago.go:285]: Out of curiosity: Will the changelog be part of the bill of materials?

**Review by saschagrunert — COMMENTED** [review]
I'm happy with that integration, well done @puerco! Maybe we can demo this one in one of the next release engineering meetings.

File: pkg/anago/anago.go (hunk 4 above)
Comment (puerco) [reply pkg/anago/anago.go:285]: Not the content of the changelog itself, but we have to list the file as an item in the file

File: pkg/anago/stage.go (hunk 1 above)
Comment (puerco) [reply pkg/anago/stage.go:635]: (yes it's just a tmp debugger I'm using to locate the artifacts : )

File: pkg/license/download.go (hunk 2 above)
Comment (puerco) [reply pkg/license/download.go:63]: Mmh I don't remember why I moved this public. Done.

File: pkg/anago/stage.go (hunk 1 above)
Comment (puerco) [reply pkg/anago/stage.go]: Done, removed

File: pkg/license/download.go (hunk 2 above)
Comment (puerco) [reply pkg/license/download.go]: I don't recall why I made this public. I've reverted the change.

**Comment by puerco** [comment]
The SBOM integration is ready to go! :tada: 

There are still details to iron out, but the big chunk is done. We are ready to produce the first set of SBOMs for `v1.22.0-alpha.3`!  :rocket: 

/hold cancel

**Comment by dims** [comment]
@puerco is the list of things in the SBOM, exactly reflect the stuff in the LICENSES/ directory in k/k? (or does it have more stuff?)

**Comment by puerco** [comment]
All the data in LICENSES should be included plus a lot more.  Most should be there now, but I noticed in the logs that some modules could not be downloaded and hence have no license declared. I have to check why. 

This PR lays the groundwork to generate the doc automatically. Once this is in, my focus will be on perfecting the output for beta.0 The licensing data of dependencies will be 1:1 to the LICENSES dir but I have yet to write some scripts to verify.

**Review by saschagrunert — APPROVED** [review]

**Comment by dims** [comment]
@puerco "plus a lot more" sounds like could be a concern...

**Pull request merged**